README.md
---
title:		"Nathan's Personal Emacs Configuration"
author:		Nathan DeGruchy
date:		2014-10-24
Time-stamp:	<2018-07-31 11:54:17 ndegruchy>
---

# Emacs Configuration

This is my personal Emacs configuration that I use across multiple
systems. I've configured it in such a way to allow me a usable
configuration that can be tweaked per-platform.

Most of it should be fairly self-explanatory, as I've taken the time
to redo some of the file layout and give them useful names.


ChangeLog
2020-05-23  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/required-packages.el (web-mode): Add modes to
	highlight on, and add some default options

2020-02-11  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/required-packages.el (ivy): Configure
	ivy-extra-directories to nil, this hides "." and ".." from ivy's
	minibuffers.
	(ivy): Move ivy-minibuffer-maps into use-package (bind-key) bind
	map

2020-02-10  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/required-packages.el, settings.d/included-packages.el:
	  Remove Ido functions, install and configure Ivy, Counsel and Swiper

2019-10-01  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/general-settings.el (battery-mode-line-format):
	Update format to hide non-functional temp gauge.

2019-09-15  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/custom-aliases.el (mpd): New alias for make-directory

2019-08-17  Nathan DeGruchy  <nathan@degruchy.org>

	* site-lisp.d/dired+.el: Add dired+

	* settings.d/: Remove custom platform settings, I don't use much
	in the way of custom settings, so strip all that out, too.

2019-06-23  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/custom-hooks.el: Add a custom hook to fix material
	theme with my emacs, as I run it in daemon mode

	* settings.d/general-settings.el (default-frame-alist): Change the
	way the font was defined

	* settings.d/linux-platform-settings.el (material-theme): Change
	to material theme to match my desktop

2019-04-08  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/linux-platform-settings.el: Remove dired-atool and
	rsync for lack of use

	* settings.d/general-settings.el: Remove font definition in the
	general settings, changed to per-platform

	* settings.d/linux-platform-settings.el: Add font for linux
	desktop

	* settings.d/macos-platform-settings.el: Remove smex keybinding

2019-03-19  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/custom-functions.el (protect-buffers): Protect
	certain buffers

2019-03-16  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/linux-platform-settings.el: Add base16 theme back in,
	cleanup font.

2019-03-15  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/required-packages.el (rotate): Add package 'rotate',
	to required packages. Allows me to rotate windows around without
	having to manually juggle buffers.

2019-03-13  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/linux-platform-settings.el (use-package emms): Re-add
	emms to emacs. I ditched mpd because it wasn't proving as useful
	to me. I use borg to organize music and so I didn't need another
	layer on top of that. Emms -> files -> play in VLC (nox)

2019-02-14  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/required-packages.el: Install Helm

2019-02-11  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/general-settings.el (delete-by-moving-to-trash): Add
	option to send files to the trash (as specified by the XDG spec)
	instead of straight-up deleting them.

2019-02-06  Nathan DeGruchy  <nathan@degruchy.org>

	* Cleanup: remove old functions and keybindings, finally, actually
	ignore the emacs-customize file.

2019-01-28  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/linux-settings.el: Move auctex to this file and put
	it behind an if statement to narrow it's install my desktop, where
	TeXLive will be installed.

	* settings.d/required-packages.el (windresize): Remove some
	packages that I don't tend to use anymore. Not that they're not
	good, I just don't find they get used.

2019-01-19  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/included-packages.el: Changed dired to use the ls
	option to group directories first in a listing. What can I say, I
	like it better.

	* ndegruchy/duplicate-line: New function duplicates the current
	line, useful if you need to make a list of directories, or other
	repetitive tasks. C-S-Backspace C-y C-y works, too, but this is
	sexier.

	* YAML-Mode: New package: YAML mode, because all the cool kids
	seem to be using YAML for configuration, so we have this.

2019-01-11  Nathan DeGruchy  <nathan@degruchy.org>

	* EMMS: Added emms to my config so I can play my music from within
	emacs. One more step to full emacs-dom!
	* PDF-tools: Added some shortcuts for annotating in PDFs in
	PDF-tools, that way I can add notes and highlights.

2018-12-22  Nathan DeGruchy  <nathan@degruchy.org>

	* Changes: Some packages removed, some added, a little
	configuration tweaks

2018-12-20  Nathan DeGruchy  <nathan@degruchy.org>

	* Minor-changes: Updated some files.

2018-12-17  Nathan DeGruchy  <nathan@degruchy.org>

	* Midnight-mode: Enable midnight mode for cleaning up buffers when
	I leave my PC on overnight. Actually runs at 04:30, in case I'm up
	late.

	* Dired-hacks: Add a number of dired-hacks files to emacs for cool
	narrowing and filtering in dired.

2018-12-08  Nathan DeGruchy  <nathan@degruchy.org>

	* custom-lisp.d/emacs-customize.el: Remove Magit packages (and
	friends) due to the built in VC being good enough for my purposes.

2018-08-07  Nathan DeGruchy  <nathan@degruchy.org>

	* settings.d/custom-keybindings.el: Add new keymap for
	`avy-goto-char`. Sometimes a word is not enough.

	* settings.d/included-packages.el: Unwrap configurations from
	`use-package`. Not really required, as they're included with
	Emacs, and it only adds overhead as it filters through the lisp
	wrapping it.

	* settings.d/custom-keybindings.el ("C-c l"): Modify keybinding to
	use new select function.

	* settings.d/custom-functions.el: Remove
	`ndegruchy/kill-whole-line`. Turns out Emacs already has a
	`kill-whole-line` function. Huh. Remove associated keybinding, too.
	(ndegruchy/select-current-line-dwim): Remove old select line
	function and replace it with this, newer, one that emulates what I
	intend to use it for.

	* settings.d/required-packages.el (magit): Remove `iedit`. I tried
	this and multiple cursors. Not appreciably better than something

lisp/functions.el
;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Functions (and advice) ;;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defun ndegruchy/c-mode-keybindings ()
  "For use in `c-mode-hook'. Unbinds some inconvenient keys that
  are set in C mode"
  (local-unset-key (kbd "C-c C-c")))

(defun ndegruchy/insert-date (format)
  "Wrapper around format-time-string"
  (interactive "MFormat: ")
  (insert (format-time-string format)))

(defun ndegruchy/cleanup-buffer ()
  "Perform a bunch of operations on the whitespace content of a
  buffer.  Including ndegruchy/indent-buffer, which should not be called
  automatically on save."
  (interactive)
  (ndegruchy/untabify-buffer)
  (delete-trailing-whitespace)
  (ndegruchy/indent-buffer))

(defun ndegruchy/untabify-buffer ()
  (interactive)
  (untabify (point-min) (point-max)))

(defun ndegruchy/indent-buffer ()
  (interactive)
  (indent-region (point-min) (point-max)))

(defun ndegruchy/kill-all-dired-buffers ()
  "Kill all dired buffers."
  (interactive)
  (save-excursion
    (let((count 0))
      (dolist(buffer (buffer-list))
        (set-buffer buffer)
        (when (equal major-mode 'dired-mode)
          (setq count (1+ count))
          (kill-buffer buffer)))
      (message "Killed %i dired buffer(s)." count ))))

(defun ndegruchy/fill-or-unfill ()
  "Like `fill-paragraph', but unfill if used twice."
  (interactive)
  (let ((fill-column
         (if (eq last-command 'ndegruchy/fill-or-unfill)
             (progn (setq this-command nil)
                    (point-max))
           fill-column)))
    (call-interactively #'fill-paragraph)))

(defun ndegruchy/unfill-paragraph ()
  "Takes a multi-line paragraph and makes it into a single line of text."
  (interactive)
  (let ((fill-column (point-max)))
    (fill-paragraph nil)))

(defun ndegruchy/kill-whole-line nil
  "kills the entire line on which the cursor is located, and
places the cursor as close to its previous position as possible."
  (interactive)
  (progn
    (let ((y (current-column))
          (a (progn (beginning-of-line) (point)))
          (b (progn (forward-line 1) (point))))
      (kill-region a b)
      (move-to-column y))))

(defun ndegruchy/select-current-word ()
"Select the word under cursor.
“word” here is considered any alphanumeric sequence with “_” or “-”."
 (interactive)
 (let (pt)
   (skip-chars-backward "-_A-Za-z0-9")
   (setq pt (point))
   (skip-chars-forward "-_A-Za-z0-9")
   (set-mark pt)))

(defun ndegruchy/select-current-line ()
  "Select the current line"
  (interactive)
  (end-of-line) ; move to end of line
  (set-mark (line-beginning-position)))

(defun ndegruchy/select-line ()
  "Select current line. If region is active, extend selection downward by line.
URL `http://ergoemacs.org/emacs/modernization_mark-word.html'
Version 2016-07-22"
  (interactive)
  (if (region-active-p)
      (progn
        (forward-line 1)
        (end-of-line))
    (ndegruchy/select-current-line)))

(defun ndegruchy/open-line-below ()
  "Inserts a line below the current line, moving the cursor to
that line and setting the indent properly"
  (interactive)
  (end-of-line)
  (newline)
  (indent-for-tab-command))

(defun ndegruchy/open-line-above ()
  "Inserts a line above the current line, moving the cursor to
that line and setting the indent properly"
  (interactive)
  (beginning-of-line)
  (newline)
  (forward-line -1)
  (indent-for-tab-command))

(defun ndegruchy/ido-choose-from-recentf ()
  "Use ido to select a recently visited file from the `recentf-list'"
  (interactive)
  (find-file (ido-completing-read "Open file: " recentf-list nil t)))

;; Create parent folder(s) when visiting a non-existant file
(defun ndegruchy/my-create-non-existent-directory ()
  (let ((parent-directory (file-name-directory buffer-file-name)))
    (when (and (not (file-exists-p parent-directory))
               (y-or-n-p (format "Directory `%s' does not exist! Create it?" parent-directory)))
      (make-directory parent-directory t))))

;; Open file in preferred app
(defun ndegruchy/open-in-external-app ()
  "Open the current file or dired marked files in external app.
The app is chosen from your OS's preference.

Version 2015-01-26
URL `http://ergoemacs.org/emacs/emacs_dired_open_file_in_ext_apps.html'"
  (interactive)
  (let* (
         (ξfile-list
          (if (string-equal major-mode "dired-mode")
              (dired-get-marked-files)
            (list (buffer-file-name))))
         (ξdo-it-p (if (<= (length ξfile-list) 5)
                       t
                     (y-or-n-p "Open more than 5 files? "))))

    (when ξdo-it-p
      (cond
       ((string-equal system-type "windows-nt")
        (mapc
         (lambda (fPath)
           (w32-shell-execute "open" (replace-regexp-in-string "/" "\\" fPath t t))) ξfile-list))
       ((string-equal system-type "darwin")
        (mapc
         (lambda (fPath) (shell-command (format "open \"%s\"" fPath)))  ξfile-list))
       ((string-equal system-type "gnu/linux")
        (mapc
         (lambda (fPath) (let ((process-connection-type nil)) (start-process "" nil "xdg-open" fPath))) ξfile-list))))))

;; Open in file manager
(defun ndegruchy/open-in-desktop ()
  "Show current file in desktop (OS's file manager)."
  (interactive)
  (cond
   ((string-equal system-type "windows-nt")
    (w32-shell-execute "explore" (replace-regexp-in-string "/" "\\" default-directory t t)))
   ((string-equal system-type "darwin") (shell-command "open ."))
   ((string-equal system-type "gnu/linux")
    (let ((process-connection-type nil)) (start-process "" nil "xdg-open" "."))
    ;; (shell-command "xdg-open .") ;; 2013-02-10 this sometimes froze emacs till the folder is closed. ⁖ with nautilus
    ) ))

(defun ndegruchy/comment-line (n)
  "Comment or uncomment current line and leave point after it.
With positive prefix, apply to N lines including current one.
With negative prefix, apply to -N lines above."
  (interactive "p")
  (comment-or-uncomment-region
   (line-beginning-position)
   (goto-char (line-end-position n)))
  (forward-line 1)
  (back-to-indentation))

(defun ndegruchy/my-asciify-string (string)
  "Convert STRING to ASCII string.
For example:
“passé” becomes “passe”"
  ;; Code originally by Teemu Likonen
  (with-temp-buffer
    (insert string)
    (call-process-region (point-min) (point-max) "iconv" t t nil "--to-code=ASCII//TRANSLIT")
    (buffer-substring-no-properties (point-min) (point-max))))

(defun ndegruchy/title-case-region-or-line (φp1 φp2)
  "Title case text between nearest brackets, or current line, or text selection.
Capitalize first letter of each word, except words like {to, of, the, a, in, or, and, …}. If a word already contains cap letters such as HTTP, URL, they are left as is.

When called in a elisp program, φp1 φp2 are region boundaries.
URL `http://ergoemacs.org/emacs/elisp_title_case_text.html'
Version 2015-04-08"
  (interactive
   (if (use-region-p)
       (list (region-beginning) (region-end))
     (let (
           ξp1
           ξp2
           (ξskipChars "^\"<>(){}[]“”‘’‹›«»「」『』【】〖〗《》〈〉〔〕"))
       (progn
         (skip-chars-backward ξskipChars (line-beginning-position))
         (setq ξp1 (point))
         (skip-chars-forward ξskipChars (line-end-position))
         (setq ξp2 (point)))
       (list ξp1 ξp2))))
  (let* (
         (ξstrPairs [
                     [" A " " a "]
                     [" And " " and "]
                     [" At " " at "]
                     [" As " " as "]
                     [" By " " by "]
                     [" Be " " be "]
                     [" Into " " into "]
                     [" In " " in "]
                     [" Is " " is "]
                     [" It " " it "]
                     [" For " " for "]
                     [" Of " " of "]
                     [" Or " " or "]
                     [" On " " on "]
                     [" The " " the "]
                     [" That " " that "]
                     [" To " " to "]
                     [" Vs " " vs "]
                     [" With " " with "]
                     [" From " " from "]
                     ["'S " "'s "]
                     ]))
    (save-restriction
      (narrow-to-region φp1 φp2)
      (upcase-initials-region (point-min) (point-max))
      (let ((case-fold-search nil))
        (mapc
         (lambda (ξx)
           (goto-char (point-min))
           (while
               (search-forward (aref ξx 0) nil t)
             (replace-match (aref ξx 1) 'FIXEDCASE 'LITERAL)))
         ξstrPairs)))))

(defun ndegruchy/select-text-in-quote ()
  "Select text between the nearest left and right delimiters.
Delimiters are paired characters: () [] {} <> «» ‹› “” ‘’ 「」 【】《》〈〉〔〕（）, including \"\".

URL `http://ergoemacs.org/emacs/modernization_mark-word.html'
version 2015-02-07
"
  (interactive)
  (let (p1 p2)
    (skip-chars-backward "^<>(“{[«‹「【《〈〔（\"‘")
    (setq p1 (point))
    (skip-chars-forward "^<>)”}]»›」】》〉〕）\"’")
    (setq p2 (point))
    (set-mark p1)))

(defun ndegruchy/select-current-line ()
  "Select current line.
URL `http://ergoemacs.org/emacs/modernization_mark-word.html'
Version 2015-02-07
"
  (interactive)
  (end-of-line)
  (set-mark (line-beginning-position)))

(defun unix-file ()
  "Change the current buffer to Latin 1 with Unix line-ends."
  (interactive)
  (set-buffer-file-coding-system 'iso-latin-1-unix t))

(defun dos-file ()
  "Change the current buffer to Latin 1 with DOS line-ends."
  (interactive)
  (set-buffer-file-coding-system 'iso-latin-1-dos t))

(defun mac-file ()
  "Change the current buffer to Latin 1 with Mac line-ends."
  (interactive)
  (set-buffer-file-coding-system 'iso-latin-1-mac t))

(defun ndegruchy/new-empty-buffer ()
  "Open a new empty buffer.
URL `http://ergoemacs.org/emacs/emacs_new_empty_buffer.html'
Version 2015-06-12"
  (interactive)
  (let ((ξbuf (generate-new-buffer "untitled")))
    (switch-to-buffer ξbuf)
    (funcall (and initial-major-mode))
    (setq buffer-offer-save t)))

(defun ndegruchy/duplicate-line()
  "Taken from `http://stackoverflow.com/a/88828'"
  (interactive)
  (move-beginning-of-line 1)
  (kill-line)
  (yank)
  (open-line 1)
  (next-line 1)
  (yank))

(defun ndegruchy/strip-smart-quotes (rStart rEnd)
  "Replace smart quotes with plain quotes in text"
  (interactive "r")
  (save-restriction (narrow-to-region rStart rEnd)
                    (goto-char (point-min))
                    (while (re-search-forward "[“”]" nil t) (replace-match "\"" nil t))
                    (goto-char (point-min))
                    (while (re-search-forward "[‘’]" nil t) (replace-match "'" nil t))))

(defun ndegruchy/rename-current-buffer-file ()
  "Renames current buffer and file it is visiting."
  (interactive)
  (let ((name (buffer-name))
        (filename (buffer-file-name)))
    (if (not (and filename (file-exists-p filename)))
        (error "Buffer '%s' is not visiting a file!" name)
      (let ((new-name (read-file-name "New name: " filename)))
        (if (get-buffer new-name)
            (error "A buffer named '%s' already exists!" new-name)
          (rename-file filename new-name 1)
          (rename-buffer new-name)
          (set-visited-file-name new-name)
          (set-buffer-modified-p nil)
          (message "File '%s' successfully renamed to '%s'"
                   name (file-name-nondirectory new-name)))))))


(defun ndegruchy/find-or-create-file-at-point ()
  "Guesses what parts of the buffer under point is a file name and opens it."
  (interactive)
  (find-file (file-name-at-point)))

(defun ndegruchy/find-or-create-file-at-point-other-window ()
  "Guesses what parts of the buffer under point is a file name and opens it."
  (interactive)
  (find-file-other-window (file-name-at-point)))

(defun file-name-at-point ()
  (save-excursion
    (let* ((file-name-regexp "[./a-zA-Z0-9\-_~]")
           (start (progn
                    (while (looking-back file-name-regexp)
                      (forward-char -1))
                    (point)))
           (end (progn
                  (while (looking-at file-name-regexp)
                    (forward-char 1))
                  (point))))
      (buffer-substring start end))))

;; after deleting a tag, indent properly
(defadvice sgml-delete-tag (after reindent activate)
  (indent-region (point-min) (point-max)))

(defadvice kill-buffer (around kill-buffer-around-advice activate)
  (let ((buffer-to-kill (ad-get-arg 0)))
    (if (equal buffer-to-kill "*scratch*")
        (bury-buffer) ad-do-it)))

(defun ndegruchy/dired-rsync (dest)
  (interactive
   (list
    (expand-file-name
     (read-file-name
      "Rsync to:"
      (dired-dwim-target-directory)))))
  ;; store all selected files into "files" list
  (let ((files (dired-get-marked-files
                nil current-prefix-arg))
        ;; the rsync command
        (tmtxt/rsync-command
         "rsync -arvz --progress "))
    ;; add all selected file names as arguments
    ;; to the rsync command
    (dolist (file files)
      (setq tmtxt/rsync-command
            (concat tmtxt/rsync-command
                    (shell-quote-argument file)
                    " ")))
    ;; append the destination
    (setq tmtxt/rsync-command
          (concat tmtxt/rsync-command
                  (shell-quote-argument dest)))
    ;; run the async shell command
    (async-shell-command tmtxt/rsync-command "*rsync*")
    ;; finally, switch to that window
    (other-window 1)))

(defun ndegruchy/move-file ()
  "Write this file to a new location, and delete the old one."
  (interactive)
  (let ((old-location (buffer-file-name)))
    (call-interactively #'write-file)
    (when old-location
      (delete-file old-location))))
